src/alarm_heap.cpp

```diff
@@ -0,0 +1,326 @@
+  alarm_info->update_alarm_in_heap(severity, alarm_pop_time);
+}
+
+void AlarmHeap::issue_alarm(const std::string& issuer,
+                            const std::string& identifier)
+{
+  unsigned int index;
+  unsigned int severity;
+  if (sscanf(identifier.c_str(), "%u.%u", &index, &severity) != 2)
+  {
+    TRC_ERROR("malformed alarm identifier: %s", identifier.c_str());
+    return;
+  }
+
+  AlarmTableDef& alarm_table_def = _alarm_table_defs->get_definition(index,
+                                                                     severity);
+
+  if (alarm_table_def.is_valid())
+  {
+    pthread_mutex_lock(&_lock);
+
+    AllAlarmsStateMap::iterator alarm = _all_alarms_state.find(index);
+
+    if (alarm != _all_alarms_state.end())
+    {
+      AlarmDef::Severity severity_as_enum = (AlarmDef::Severity)severity;
+      AlarmSeverityChange severity_change =
+                  (*alarm).second->new_alarm_severity(severity_as_enum);
+
+      if (severity_change == AlarmSeverityChange::INCREASED_SEVERITY)
+      {
+        TRC_DEBUG("Severity of the alarm %lu has increased", index);
+        update_alarm_in_heap((*alarm).second, severity_as_enum, ALARM_INCREASED);
+        _cond->signal();
+      }
+      else if (severity_change == AlarmSeverityChange::REDUCED_SEVERITY)
+      {
+        TRC_DEBUG("Severity of the alarm %lu has reduced", index);
+        _alarm_heap.insert((*alarm).second->alarm_in_heap());
+        update_alarm_in_heap((*alarm).second, severity_as_enum, ALARM_REDUCED);
+        _cond->signal();
+      }
+      else
+      {
+        // The severity of the alarm hasn't changed. Remove it from the heap
+        // (safe to call even if it's already not in the heap).
+        _alarm_heap.remove((*alarm).second->alarm_in_heap());
+      }
+    }
+
+    pthread_mutex_unlock(&_lock);
+  }
+  else
+  {
+    TRC_ERROR("Unknown alarm definition: %s", identifier.c_str());
+  }
+}
+
+void AlarmHeap::sync_alarms()
+{
+  TRC_STATUS("Resyncing all alarms");
+
+  pthread_mutex_lock(&_lock);
+
+  for (AllAlarmsStateMap::iterator it = _all_alarms_state.begin();
+                               it != _all_alarms_state.end();
+                               it++)
+  {
+    AlarmDef::Severity current_severity = (*it).second->get_severity();
+
+    if (current_severity != AlarmDef::Severity::UNDEFINED_SEVERITY)
+    {
+      update_alarm_in_heap((*it).second, current_severity, ALARM_RESYNC_DELAY);
+    }
+  }
+
+  _cond->signal();
+  pthread_mutex_unlock(&_lock);
+}
+
+void AlarmHeap::handle_failed_alarm(AlarmTableDef& alarm_table_def)
+{
+  TRC_DEBUG("Handling a failed callback for the alarm: %lu",
+            alarm_table_def.alarm_index());
+
+  pthread_mutex_lock(&_lock);
+
+  AllAlarmsStateMap::iterator alarm =
+                          _all_alarms_state.find(alarm_table_def.alarm_index());
+
+  if (alarm != _all_alarms_state.end())
+  {
+    if ((*alarm).second->should_resend_alarm(alarm_table_def.severity()))
```

Comment:
If you renamed `should_resend_alarm` to `is_severity_up_to_date`, would this be clearer?